src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/CodeBasedDependencyAlgorithm.cs

```diff
@@ -118,23 +120,77 @@ public static void AddDependenciesDueToMethodCodePresence(ref DependencyList dep
             }
         }
 
-        private static void AddPInvokeParameterDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc parameter)
+        public static void AddInterestingInteropConstructedTypeDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
         {
-            if (parameter.IsDelegate)
+            if (type.IsDelegate)
             {
-                dependencies.Add(factory.NecessaryTypeSymbol(parameter), "Delegate Marshalling Stub");
+                var delegateType = type as MetadataType;
+                if (delegateType != null && delegateType.HasCustomAttribute("System.Runtime.InteropServices", "UnmanagedFunctionPointerAttribute"))
+                {
+                    AddDependenciesDuePInvokeDelegate(ref dependencies, factory, delegateType);
+                }
+            }
+            else if (type.IsValueType && type.IsTypeDefinition && !(type is NativeStructType))
+            {
+                var structType = type as MetadataType;
+                if (structType != null && structType.HasCustomAttribute("System.Runtime.InteropServices", "StructLayoutAttribute"))
+                {
+                    AddDependenciesDuePInvokeStruct(ref dependencies, factory, type);
+                }
+            }
+        }
 
-                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetOpenStaticDelegateMarshallingStub(parameter)), "Delegate Marshalling Stub");
-                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetClosedDelegateMarshallingStub(parameter)), "Delegate Marshalling Stub");
-                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetForwardDelegateCreationStub(parameter)), "Delegate Marshalling Stub");
+
+        /// <summary>
+        /// For Marshal generic APIs(eg. Marshal.StructureToPtr<T>, GetFunctionPointerForDelegate) we add
+        /// the generic parameter as dependencies so that we can generate runtime data for them
+        /// </summary>
+        public static void AddMarshalAPIsGenericDependencies(ref DependencyList dependencies, NodeFactory factory, MethodDesc method)
+        {
+            TypeDesc owningType = method.OwningType;
+            MetadataType metadataType = owningType as MetadataType;
+
+            if (metadataType != null)
+            {
+                if (metadataType.Name == "Marshal" && metadataType.Namespace == "System.Runtime.InteropServices")
```

Comment:
Do we need some sort of method name check? Does e.g. `Marshal.UnsafeAddrOfPinnedArrayElement<T>` really need to go through this?